src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs

```diff
@@ -927,6 +927,108 @@ public static class PipelineStageDefinitionBuilder
             return Match(new ExpressionFilterDefinition<TInput>(filter));
         }
 
+        /// <summary>
+        /// Creates a $merge stage.
+        /// </summary>
+        /// <typeparam name="TInput">The type of the input documents.</typeparam>
+        /// <typeparam name="TOutput">The type of the output documents.</typeparam>
+        /// <param name="outputCollection">The output collection.</param>
+        /// <param name="mergeOptions">The merge options.</param>
+        /// <returns>The stage.</returns>
+        public static PipelineStageDefinition<TInput, TOutput> Merge<TInput, TOutput>(
+            IMongoCollection<TOutput> outputCollection,
+            MergeStageOptions<TOutput> mergeOptions)
+        {
+            Ensure.IsNotNull(outputCollection, nameof(outputCollection));
+            Ensure.IsNotNull(mergeOptions, nameof(mergeOptions));
+
+            if (mergeOptions.LetVariables != null && mergeOptions.WhenMatched != MergeStageWhenMatched.Pipeline)
```

Comment:
Good point.

@jyemin what do you think?

Reply:
I am fine with builders making these sorts of assertions.

Reply:
@jyemin , I'm fine too but maybe it will be better if we are consistent in such validation logic? Which behavior is recommended for this case in general? If we don't have such rules, maybe it's better to formulate them?
It can wait for a next ticket with this situation :)

Reply:
I don't know what the convention is in the .NET driver.  I'm fine with following it, whatever it is.